proposals/0278-screenmanager-layout-management.md

```diff
@@ -11,33 +11,96 @@ Adds a screen manager method for changing the current layout to a new layout.
 ## Motivation
 The manager layer should be able to handle most tasks that used to be handled by RPCs, and the Screen Manager in particular should be able to handle manipulating the screen and templates without requiring the developer to fall back to RPCs in any cases.
 
-One big area that it is currently missing is the ability to manage changing templates, currently done with the `SetDisplayLayout` RPC (on RPC spec <6.0) and `Show` RPC (on RPC spec >6.0). This proposal adds a way to change the layout while taking care of backward compatibility concerns.
+One big area that it is currently missing is the ability to manage changing templates, currently done with the `SetDisplayLayout` RPC (on RPC spec <6.0) and `Show` RPC (on RPC spec >=6.0). This proposal adds a way to change the layout while taking care of backward compatibility concerns.
 
 ## Proposed solution
-We will add a new screen manager method for changing the current layout to a new layout.
+We will add a new screen manager method for changing the current layout to a new layout. This also includes updating the template's data immediately on v6.0+ and sequentially on <v6.0, as well as the template's color schemes on RPC v5.0+ systems.
 
 ### iOS APIs
 ```objc
 @interface SDLScreenManager : NSObject
 
-/// Change the current layout to a new layout. You can get a list of SDL-defined layout strings in SDLPredefinedLayout.
-- (void)changeLayout:(NSString *)newLayout withCompletionHandler:(SDLScreenManagerUpdateCompletionHandler)handler;
+/// Change the current layout to a new layout, update the layout's night and day color schemes and update the text, graphics, buttons and template title.
+- (void)changeLayout:(SDLTemplateLayout *)templateUpdates withCompletionHandler:(nullable SDLScreenManagerUpdateCompletionHandler)handler;
```

Comment:
I'm just thinking through this. If they provide a textField 1 & 2 only, would that clear everything else, or leave everything else alone (like happens currently in Screen Manager T&G updates)? I think that we'd want to do the second for now. If you agree, I think we should make that clear in the docs.

Reply:
You raised an interesting question. Since the values set for the text, graphics, buttons and template title persist between layout changes, if only the values set in the`templateUpdates` change then how can we keep track of whether the developer wants to leave the value unchanged or clear the value. Does setting `textField1` to `nil` mean we should clear the value or that we should not update the value at all? We could say that setting an empty string clears the `textField1` but I am not sure what the clear value would be for `primaryGraphic`, which has a data type of `SDLArtwork`.

Reply:
Honestly, the easiest way to do that would be to create a new (class) initializer for `SDLArtwork` like `[SDLArtwork clear]` or `nothing`. Otherwise, I don't know how we'd do that. You could also add a `SDLTemplateConfiguration` option to `clearUnsetData`.

Another option I just thought of that might be helpful is to remove all the fields from the `SDLTemplateLayout` class and then instead use `SDLTemplateConfiguration` which already exists. Then, if the dev batches the update, it'll all be merged together or set separately. If they don't batch, then it'll be set separately.

Reply:
I think I like that latter option quite a bit.

Reply:
I like your solution. Just to make sure we are on the same page, it sounds like you want to remove `SDLTemplateLayout` and just use `SDLTemplateConfiguration` like so:

```
- (void)changeLayout:(SDLTemplateConfiguration *)templateConfiguration withCompletionHandler:(nullable SDLScreenManagerUpdateCompletionHandler)handler;
```

The developer would do something like the following to batch the updates?
```
let templateConfiguration = SDLTemplateConfiguration(predefinedLayout: .graphicWithText)
templateConfiguration.dayColorScheme = SDLTemplateColorScheme(primaryColor: .red, secondaryColor: .blue, backgroundColor: .gray)

sdlManager.screenManager.beginUpdates()
sdlManager.screenManager.changeLayout(templateConfiguration) { error in
    // The template updated
}
sdlManager.screenManager.textField1 = "new Template set" // This will updat the current text in textField1
sdlManager.screenManager.textField2 = nil // This will clear the current text in textField2
sdlManager.screenManager.endUpdates { error in
    // Text, graphics, and buttons were updated
}
```

Reply:
If the `Show` request fails because the template is unsupported, then the entire `Show` request will be rejected (i.e. text, graphics and buttons will not be updated) however if the `SetDisplayLayout` request fails then the subsequent `Show` request will likely still succeed. Should it be the default to send another `Show` request without the template information if the request fails due to an invalid template (not even sure if we would be able to parse the error for this information) or should we let developers set the behavior?

Reply:
No, if the developer is changing the text / graphics alongside the layout change, I think it's reasonable to not set that text / graphics if the layout fails to change. If we're using the `SetDisplayLayout` and the layout change fails, we would not continue with the sequential `Show`. The same would happen with a `Show` with a layout change and the sequential soft buttons.